Fix redefinition of non-GDI Windows basic types in win32structs.h

hughbe opened the pull request:
This causes some of the unit tests to fail to compile on Windows:

> c:\users\hugh\documents\github\libgdiplus\src\win32structs.h(129): error C2371: 'TCHAR': redefinition; different basic types
> c:\program files (x86)\windows kits\10\include\10.0.15063.0\ucrt\tchar.h(180): note: see declaration of 'TCHAR'

dnfclas commented:
@hughbe,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

hughbe commented:
I've updated the PR to fix duplicate non-GDI definitions that are already supplied from Windows.h.

This also fixed a source compat issue with GDI+: it takes a `wchar_t` on Windows with GDI+, but an `guint16` in libgdiplus.

However, this now causes the build on Windows to fail, due to:

```
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(25): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(25): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(71): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(71): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(119): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(119): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(168): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(168): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(220): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(220): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(274): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(274): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(328): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(328): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(410): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(410): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(481): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(481): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(636): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(636): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(693): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(693): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(762): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(762): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(801): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(801): note: Types pointed to are unrelated; conversion requires reinterpret_cast, C-style cast or function-style cast
1>c:\users\hugh\documents\visual studio 2017\projects\consoleapplication1\consoleapplication1\testgraphics.cpp(840): error C2664: 'GpStatus GdipLoadImageFromFile(const WCHAR *,GpImage **)': cannot convert argument 1 from 'gunichar2 *' to 'const WCHAR *'
```

I supposed one course of action could be to unconditionally define `WCHAR` as `wchar_t` on non-Windows platforms. This would fix the source compat issues against GDI+ on Windows.

E.g. the following code works on Windows. Before this PR, it failed to compile against libgdiplus on Windows, but it now does. However, it still fails to compile against libgdiplus on non-Windows platforms.

```
static void test_privateFontCollectionAddFontFile()
{
	GpStatus status;
	GpFontCollection *collection;
	GdipNewPrivateFontCollection (&collection);

	const WCHAR *directory = L"C:\\Users\\hugh\\Documents\\Visual Studio 2017\\Projects\\ConsoleApplication1\\ConsoleApplication1";
	const WCHAR *filename = L"C:\\Users\\hugh\\Documents\\Visual Studio 2017\\Projects\\ConsoleApplication1\\ConsoleApplication1\\CodeNewRoman.ttf";

	status = GdipPrivateAddFontFile (NULL, filename);
	assert (status == InvalidParameter);

	status = GdipPrivateAddFontFile (collection, NULL);
	assert (status == InvalidParameter);

	status = GdipPrivateAddFontFile (collection, L"");
	assert (status == FileNotFound);

	status = GdipPrivateAddFontFile (collection, L"NoSuchFile");
	assert (status == FileNotFound);

	status = GdipPrivateAddFontFile (collection, directory);
	assert (status == GenericError);

	status = GdipPrivateAddFontFile (collection, filename);
	assert (status == Ok);

	GdipDeletePrivateFontCollection (&collection);
}
```

@qmfrederik this may need some more discussion - let me know what you think.

akoeplinger commented on src/win32structs.h:
```diff
@@ -83,15 +83,21 @@
+typedef guint32 ARGB;
+typedef float REAL;
+
+#if defined(_WIN32)
```
any reason for _WIN32 vs WIN32 which is used in other parts of the codebase?

hughbe replied:
Force of habit - changed these to `WIN32`

akoeplinger commented on src/win32structs.h:
```diff
@@ -83,15 +83,21 @@
+#include <Windows.h>
+#else
+
 typedef int LANGID;
```
unrelated to the PR but according to https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751(v=vs.85).aspx LANGID should be WORD (16-bit unsigned integer), defining it as int looks like a bug to me?

hughbe replied:
Indeed. I'm not sure what the impact of the bug was or how it would manifest itself (maybe binary compatability?), but I fixed it anyway

hughbe commented on src/win32structs.h:
```diff
@@ -83,15 +83,21 @@
+
+#if defined(_WIN32)
+#define NOGDI
+#include <Windows.h>
```
Interesting - I'll be sure to use lowercase in my own projects too!

hughbe commented:
@akoeplinger I've addressed your PR feedback, let me know if you need anything else

qmfrederik commented:
No real comments from my side, sorry for the delay, had some other things to attend to (and I'll be off for the coming weeks).

If it makes the unit tests work on both Unix & Windows, then LTGM :).

Do you have a `vcxproj` file for the tests? The tests do not seem to run as part of the Windows Jenkins build, correct?

hughbe commented:
@qmfrederik I did create one but it's ugly right now and I haven't had time to clean it up.

The problem is that each test file has it's own main method, so I'm not sure what the best way to do this is - we get compilation errors if they are all together in the same project

qmfrederik commented:
@hughbe Got it. Yeah, you can either create a `.vcxproj` file for each test - that would be closes to what's done in the Unix world.

Alternatively, you could try something like having one test executable, where you pass the name of the test as a command line argument.

For each test, you could rename the `main` method if being compiled as part of a `vcxprpoj` - for example by detecting `_MSC_VER` or a preprocessor variable you define on your own.

```
int
#if _MSC_VER
test_name_main (int argc, char **argv)
#else
main (int argc, char **argv)
{
#endif
```

It's still not ideal but may be a starting point.

hughbe commented:
Okay. I'll take a look at that in a future PR. Would something like googletest be a good idea?\

That aside, is there anything blocking this PR?

akoeplinger commented:
> That aside, is there anything blocking this PR?

Nothing, I just got confused about what exactly this breaks from the comment in https://github.com/mono/libgdiplus/pull/73#issuecomment-312475633, but now I see it just makes things better. Thanks 👍
